torch/fx/graph_module.py

```diff
@@ -296,18 +301,65 @@ def recompile(self) -> None:
         cls = type(self)
         cls.forward = _forward_from_src(self._code)
 
-        cls_call = cls.__call__
+        # If we're calling `recompile` on a GraphModule that has already
+        # been recompiled, `cls.__call__` would be `wrapped_call`.
+        # We don't want to wrap `wrapped_call` again (as this would
+        # ultimately lead to O(num_recompiles) "unwrappings" later), so
+        # we use the instance attribute `_cls_call` to store the
+        # original implementation of `__call__`
+        if not self._cls_call:
+            self._cls_call = cls.__call__
 
         def print_full_traceback(exctype, value, tb):
             traceback.print_exception(exctype, value, tb)
 
+        # Previously, if an error occurred when valid
+        # symbolically-traced code was run with an invalid input, the
+        # user would see the source of the error as coming from
+        # `File "<eval_with_key_N">`, where N is the number of times
+        # `recompile` had been called on the GraphModule. We use this
+        # function to generate a more informative error message. We
+        # return the traceback itself, a message explaining that the
+        # error occurred in a traced Module's generated forward
+        # function, and five lines of context surrounding the faulty
+        # faulty line
+        def generate_error_message(tb: TracebackType) -> str:
+            # auxiliary variables (for readability)
+            frame_summary = traceback.extract_tb(tb)[-1]
```

Comment:
This is implicitly assuming that the error is occurring in the `GraphModule` code itself, but in reality an exception propagating through here could come from any transitive call. For example:
```
import torch
import torch.fx

class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.linear = torch.nn.Linear(3, 4)

    def forward(self, x):
        return self.linear(x)

f = torch.fx.symbolic_trace(Foo())
f(torch.rand(5, 5))
"""
Traceback (most recent call last):
  File "/data/users/jamesreed/pytorch/torch/fx/graph_module.py", line 350, in wrapped_call
    return self._cls_call(self, *args, **kwargs)
  File "/data/users/jamesreed/pytorch/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "<eval_with_key_0>", line 3, in forward
    linear_1 = self.linear(x);  x = None
  File "/data/users/jamesreed/pytorch/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/data/users/jamesreed/pytorch/torch/nn/modules/linear.py", line 94, in forward
    return F.linear(input, self.weight, self.bias)
  File "/data/users/jamesreed/pytorch/torch/nn/functional.py", line 1755, in linear
    ret = torch.addmm(bias, input, weight.t())
RuntimeError: mat1 and mat2 shapes cannot be multiplied (5x5 and 3x4)
Call to “exec” using an FX-traced Module, line 1755 of the traced Module’s generated forward function:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
Traceback (most recent call last):
  File "relu_test.py", line 13, in <module>
    f(torch.rand(5, 5))
  File "/data/users/jamesreed/pytorch/torch/fx/graph_module.py", line 358, in wrapped_call
    raise e.with_traceback(None)
RuntimeError: mat1 and mat2 shapes cannot be multiplied (5x5 and 3x4)
"""
```

In this case the extra context is actually incorrect and probably confusing. I think maybe at 351 you could peel off the top frame from the traceback and check if that filename contains `eval_with_cache`, then dispatch into `generate_error_message`, otherwise just rethrow

Reply:
Can we pass in `topmost_framesummary` here rather than doing `extract_tb` again?